vendor/parser-extended/ast_mapper.ml

```diff
@@ -485,6 +487,31 @@ module M = struct
     | Pstr_attribute x -> attribute ~loc (sub.attribute sub x)
 end
 
+module Comprehension = struct
+  (* List/array comprehensions *)
+
+  let map_iterator sub = function
+    | Range { start; stop; direction } ->
+      Range
+        { start = sub.expr sub start
+        ; stop = sub.expr sub stop
+        ; direction = Flag.map_direction sub direction }
+    | In e -> In (sub.expr sub e)
+
+  let map_clause_binding sub {pattern; iterator; attributes} =
+    { pattern = sub.pat sub pattern
+    ; iterator = map_iterator sub iterator
+    ; attributes = sub.attributes sub attributes }
+
+  let map_clause sub = function
+    | For cbl -> For (List.map (map_clause_binding sub) cbl)
+    | When e -> When (sub.expr sub e)
+
+  let map sub {comp_body; clauses} =
```

Comment:
Comprehensions are unusual in that they might be the only syntactic form in OCaml that bind a thing *after* it's in scope. Is there a place where an ast_mapper assumes that binding sites precede usage sites? Maybe -- so it's probably best to reorder these.

Reply:
Any `ast_mapper` that depends on the order that things get mapped certainly should not be doing that. In any case, this code is directly copied from the way the compiler does it (see line 551 of `parser-standard/ast_mapper.ml`)